### DISCUSSION TITLE
Page.insert_box：If the text has too many characters, it will not be displayed completely.

### DISCUSSION THREAD

**Discussion by LittleMoDel** [pymupdf/PyMuPDF#3113] [Q&A] [answered]
### Description of the bug

t1 = "text with rotate = 0.can't see me"，Can't display...

```python
import fitz

doc = fitz.open()  # new or existing PDF
page = doc.new_page()  # new page, or choose doc[n]

# write in this overall area
rect = fitz.Rect(100, 100, 300, 150)

# partition the area in 4 equal sub-rectangles
CELLS = fitz.make_table(rect, cols=4, rows=1)

t1 = "text with rotate = 0.can't see me"  # these texts we will written
t2 = "text with rotate = 90."
t3 = "text with rotate = 180."
t4 = "text with rotate = 270."
text = [t1, t2, t3, t4]
red = fitz.pdfcolor["red"]  # some colors
gold = fitz.pdfcolor["gold"]
blue = fitz.pdfcolor["blue"]

shape = page.new_shape()  # create Shape
for i in range(len(CELLS[0])):
    shape.draw_rect(CELLS[0][i])  # draw rectangle
    shape.insert_textbox(
        CELLS[0][i], text[i], fontname="hebo", color=blue, rotate=90 * i
    )

shape.finish(width=0.3, color=red, fill=gold)
shape.commit()  # write all stuff to the page
doc.save("D:/tst/xxx.pdf")
```

### How to reproduce the bug

![image](https://github.com/pymupdf/PyMuPDF/assets/34939400/5c695f30-ba8f-4395-b8a0-5eec38098780)

### PyMuPDF version

1.23.19

### Operating system

Windows

### Python version

3.11

**Comment by JorjMcKie** [comment]
This is **_normal!_**
And also documented.

If you do not want to care about whether something fits or not, use `page.insert_htmlbox` instead.

**Comment by LittleMoDel** [comment]
`page.insert_htmlbox`，and `doc.subset_fonts()` like this:

```python
import fitz_new as fitz

# Example text
text = """
    <p>Hello World!</p>
    <p class="a">你好世界！</p>
    <span class="b">你好世界！我是span！</span>
    """
arch = fitz.Archive("D:/")
# These statements define which font file to use for regular, bold,
# italic and bold-italic text.
# We assign an arbitary common font-family for all 4 font files.
# The Story algorithm will select the right file as required.
# We request to use "comic" throughout the text.
css = """
@font-face {font-family: 'SimSun-ExtB Regular'; src: url(simsunb_reduce.ttf);}
@font-face {font-family: 'Arial Regular'; src: url(arial_reduce.ttf);}
@font-face {font-family: 'XinYuGongZhangJiaSongA'; src: url(OPPOSans.ttf);}
* {font-family: 'Arial Regular';}
.a {font-family: 'XinYuGongZhangJiaSongA';}
.b {font-family: 'SimSun-ExtB Regular';}
"""

doc = fitz.Document()
page = doc.new_page(width=300, height=300)  # make small page

page.insert_htmlbox(page.rect, text, css=css, archive=arch)

doc.subset_fonts(verbose=True)  # build subset fonts to reduce file size
doc.save("D:/123_new.pdf")
```
![image](https://github.com/pymupdf/PyMuPDF/assets/34939400/f19244e1-56cc-48e3-8233-a3b4e56535b9)

![image](https://github.com/pymupdf/PyMuPDF/assets/34939400/f2cf6f77-c906-459e-b7af-e6ae26cf6be0)


Is this normal？

**Comment by JorjMcKie** [comment]
Yes - set verbose to False if you don't want to see the messages.

Your save statement however is too careless!
Subset font building leaves the old large font there - you must compress!!!
`doc.ez_save(...)` has the necessary options.

**Comment by LittleMoDel** [comment]
`doc.ez_save(...)`  Thank you very much!

'NotoSerif-Regular' and 'Droid Sans Fallback Regular',I did not set these two font.
I figure out the `"SimSun-ExtB Regular"` not support chinese,so that replaced by 'Droid Sans Fallback Regular'!
but `'NotoSerif-Regular' `, I can‘t undestand...

**Comment by JorjMcKie** [comment]
> but `NotoSerif-Regular` , I can't undestand...

This is one of MuPDF's default fonts when no other font-family name is applicable. Depending on the Unicodes encountered, either "NotoSerif-Regular" is chosen or "CharisSIL-Serif..." in such a case.
You must be using a Unicode that is not contained in any of the given fonts ...

But you can find out yourself what has happened: extract the text via "dict" or "rawdict" (respectively "json"/"rawjson") and you will which font has been used for what.

**Comment by LittleMoDel** [comment]
I have something can't handle about `page.insert_htmlbox`, code :

```python
import fitz_new as fitz
doc = fitz.open("C:/tst/old.pdf") # open document
page = doc[0] # get the 1st page of the document
rect2=fitz.Rect(100,156,220,170)

text = """<div class="cs">分切尺寸：200*300mm</div>"""
arch = fitz.Archive("D:/")
CSS = """
@font-face {font-family: 'simsun'; src: url(simsun.ttc);}
@font-face {font-family: 'arial'; src: url(arial.ttf);}
# @font-face {font-family: 'OPPOSans'; src: url(OPPOSans.ttf);}

* {
    font-family: 'simsun',sans-serif;
    font-size:7.98pt;
    color:red;
    margin:0px;
    padding:0px;
    border-width:0px;
}
.cs {
    font-family: arial,sans-serif;
    # font-size:7.98pt;
}
"""
page.insert_htmlbox(
    rect2, 
    text, 
    css=CSS, 
    archive=arch,
    rotate=0,
    )
print(page.get_text("dict", sort=False))
doc.subset_fonts(verbose=True)
doc.ez_save("C:/tst/new.pdf")
doc.close()
```
---
**I want the text start at x0=100, but I  runned the code above ,It gives me** 

`[{'size': 7.980000019073486, 'flags': 4, 'font': 'SimSun', 'color': 16711680, 'ascender': 0.859375, 'descender': -0.140625, 'text': '分切尺寸：200*300mm', 'origin': (101.0, 164.1820068359375), 'bbox': (101.0, 157.32418823242188, 176.8100128173828, 165.30419921875)}]`

**The text just starts at the 101 !**


**And I changed html tag:** `text = """<div class="cs">分切尺寸：200*300mm</div>"""` to `text = """<p class="cs">分切尺寸：200*300mm</p>"""`, 

**It gives me another position,and the fontsize=6.359182357788086.....:**

`[{'size': 6.359182357788086, 'flags': 4, 'font': 'SimSun', 'color': 16711680, 'ascender': 0.859375, 'descender': -0.140625, 'text': '分切尺寸：200*300mm', 'origin': (100.79689025878906, 168.0823974609375), 'bbox': (100.79689025878906, 162.6174774169922, 161.20912170410156, 168.97665405273438)}]`


[new.pdf](https://github.com/pymupdf/PyMuPDF/files/14094609/new.pdf)
[old.pdf](https://github.com/pymupdf/PyMuPDF/files/14094610/old.pdf)

**Reply by JorjMcKie** [reply, marked as answer]
> The text just starts at the 101 !

The method makes an internal default decision about the margin: `"body {margin:1px;}" + css`, where `css` is your parameter in the method.
So overwrite with desired margin value.

**Reply by LittleMoDel** [reply]
I set `body {margin:0px;}` and solved.
Thank you so much for your help. I fully appreciate it.

**Reply by JorjMcKie** [reply]
Very pleased you like this feature!
It is worthwhile to look at all the examples in the documentation. The method can be used for East-Asian scripts like Devanagari, Tamil, Bengali, etc. In all cases, fonts are automatically pulled in as required and the advanced **_text shaping library_** HarfBuzz takes care of using the correct ligatures.

Another application is inserting images with transparency - even when the image itself is not transparent.
This is something causing a large effort normally.
